docs/cookbook/langgraph_vs_mirascope/quickstart.md

```diff
@@ -0,0 +1,390 @@
+
+    query: str = Field(..., description="The users question.")
+
+    def call(self) -> str:
+        """Prompts a human to enter a response."""
+        print(f"The AI has requested assistance. Here is the question: {self.query}")
+        response = input("(Human): ")
+        return f"Human response: {response}"
+
+class Chatbot(BaseModel):
+    ...
+
+    @openai.call(model="gpt-4o-mini", stream=True, tools=[WebSearch, RequestAssistance])
+    @prompt_template(
+        """
+        SYSTEM:
+        You are an expert web searcher. 
+        Your task is to answer the user's question using the provided tools.
+        You have access to the following tools:
+            - `WebSearch`: Search the web for information.
+            - `RequestAssistance`: Request assistance from a human expert if you do not
+                know how to answer the question.
+
+        Once you have gathered all of the information you need, generate a writeup that
+        strikes the right balance between brevity and completeness. The goal is to
+        provide as much information to the writer as possible without overwhelming them.
+
+        MESSAGES:
+        {self.history}
+
+        USER:
+        {question}
+        """
+    )
+    def _call(self, question: str | None = None): ...
+
+    ...
+
+Chatbot.run()
+# (User): What is my name?
+# (Assistant): The AI has requested assistance. Here is the question: What is my name?
+# (Human): The users name is Will.
+```
+
+## Part 4: Time Travel also known as list splicing
+
+In order to revisit previous states, you can take the history list that stores all the messages and manipulate it however you want by basic list splicing.
+For example, if you want to "rewind" your state, you can simply remove user and assistant messages and then run your chatbot.
+
+```python
+chatbot = ChatBot()
+chatbot.run()
+# (User): Hi there! My name is Will.
+# (Assistant): It's nice to meet you, Will! I'm an AI assistant created by Anthropic. I'm here to help you with any questions or tasks you may have. Please let me know how I can assist you today.
+
+
+chatbot.run()
+# (User): Remember my name?
+# (Assistant): Of course, your name is Will. It's nice to meet you again!
+
+chatbot.history = chatbot.history[:-4]
+# (User): Remember my name?
+# (Assistant): I'm afraid I don't actually have the capability to remember your name. As an AI assistant, I don't have a persistent memory of our previous conversations or interactions. I respond based on the current context provided to me. Could you please restate your name or provide more information so I can try to assist you?
+```
+
+While in this example, we are calling the run function multiple times, there is nothing stopping you from updating the history inside the Chatbot. Note that in real-world applications, conversation history is typically stored in a cache, database, or other persistent storage systems, so add `computed_fields` as necessary to retrieve from storage.
```

Comment:
maybe worth adding a little blurb here mentioning the use of something like RAG to pull and insert only relevant history?